core/err.go

```diff
@@ -0,0 +1,91 @@
+package core
+
+const (
+	errCodeNoErr = iota
+	errCodeBufferExpired
+	errCodeBufferTooSmall
+	errCodeCofferExpired
+	errCodeDecryptionFailed
+	errCodeInvalidKeyLength
+	errCodeNullBuffer
+	errCodeNullEnclave
+
+	// Error string context.
+	ctx = "memguard/core: "
+)
+
+type errCode uint
+
+type memguardErr struct {
+	s string
+	c errCode
+}
+
+func (e memguardErr) Error() string {
+	return e.s
+}
+
+var errors = [...]memguardErr {
+	{ctx + "no error", errCodeNoErr},
+	{ctx + "buffer has been purged from memory and can no longer be used", errCodeBufferExpired},
+	{ctx + "the given buffer is too small to hold the plaintext", errCodeBufferTooSmall},
+	{ctx + "attempted usage of destroyed key object", errCodeCofferExpired},
+	{ctx + "decryption failed", errCodeDecryptionFailed},
+	{ctx + "key must be exactly 32 bytes", errCodeInvalidKeyLength},
+	{ctx + "buffer size must be greater than zero", errCodeNullBuffer},
+	{ctx + "enclave size must be greater than zero", errCodeNullEnclave},
+}
+
+func isXError(e error, c errCode) bool {
+	ei, ok := e.(memguardErr)
+	return ok && ei.c == c
+}
+
+// IsBufferExpired returns a boolean indicating whether the error is
+// known to report that an operation was attempted on or with a buffer
+// that has been destroyed.
+func IsBufferExpired(e error) bool {
```

Comment:
Is it possible to have these as methods instead?

Reply:
see https://github.com/awnumar/memguard/issues/111#issuecomment-517474611